diff --git a/plugins/media-keys/gsd-media-keys-manager.c b/plugins/media-keys/gsd-media-keys-manager.c
index 42fa14da..d3896746 100644
--- a/plugins/media-keys/gsd-media-keys-manager.c
+++ b/plugins/media-keys/gsd-media-keys-manager.c
@@ -823,7 +823,7 @@ media_key_new_for_path (GsdMediaKeysManager *manager,
 
         key = media_key_new ();
         key->key_type = CUSTOM_KEY;
-        key->modes = GSD_ACTION_MODE_LAUNCHER;
+        key->modes = GSD_ACTION_MODE_LAUNCHER | SHELL_ACTION_MODE_POPUP;
         key->custom_path = g_strdup (path);
         key->custom_command = command;
         key->grab_flags = META_KEY_BINDING_NONE;
